### DISCUSSION TITLE
updateQuery in React & TypeScript

### DISCUSSION THREAD

**Discussion by mariusflorescu** [urql-graphql/urql#1414] [Q&A] [answered]
Hello. I'm currently experiencing some errors when trying to do an update query in react and typescript.
I've tried several examples found on the issues/discussion tab + the documentation but nothing really works for me.

That being said, let's dive into the code.

```
export const GET_CARS_QUERY = `
  query fetchAllCars {
    fetchCars{
      id
      numberplate
      data1
      data2
      data3
    }
  }
`;
```

```
const cache = cacheExchange({
  updates:{
    Mutation:{
      addCar: (result,args,cache,info) => {
        cache.updateQuery({query:GET_CARS_QUERY},data => {
            //
            return data;
        })
      }
    }
  }
})
```

Where '//' are i tried things like:

data.fetchCars.push -> Property 'push' does not exist on type 'string | number | boolean | Data | ScalarObject | Scalar[] | NullArray<Data>'.
  Property 'push' does not exist on type 'string'.

Console logging the data will show that fetchCars is an array of objects, coressponding to my schema.

I've also tried:

```
return {
    ...data,
    result
}
```

will throw the following error (it is highlighed at the parameter data):

Argument of type '(data: Data | null) => { result: Data; __typename?: string; _id?: string | number | null | undefined; id?: string | number | null | undefined; }' is not assignable to parameter of type '(data: Data | null).

I do mention that all my mutation / queries work just fine. The only problem is with the updateCache.

Thank you a lot!

**Comment by kitten** [comment, marked as answer]
That's just an error due to some typings in our TypeScript bindings; to put it differently, at least for now, we don't have type generation for any part of Graphcache (pending #901). That means that any given field may have a lot of different types for their values. In this case what you're seeing is that your field can have one of many different values.

So either you have to type these as `any`, do some preliminary type checks, or otherwise update it so that `data.fetchCars` is recognised by TS to be an array.

The simplest solution is to write `(data.fetchCars as any[])` and the most accurate solution would be:

```ts
import { Data } from '@urql/exchange-graphcache';

(data.fetchCars as Data[])
```

On a side note, if this is your GraphQL schema, I'd advise you to avoid words like `fetch` in query that make these fields sound imperative and instead opt for non-imperative sounding fields for query fields (unlike mutations) like `cars` rather than `fetchCars` ✌️

**Reply by mariusflorescu** [reply]
Thank you for your reply.
I'm sorry, but I do not understand exaclty what do you mean by: otherwise update it so that data.fetchCars is recognised by TS to be an array.

Where I do the query, 'fetchCars' is recognized as an array,.

```
interface Cars {
  fetchCars: Car[];
}
```

Thank you again✌️

**Reply by mariusflorescu** [reply]
Thank you for your useful edit. 
I managed to do it with the solution you provided.
✌️
